meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,435 @@
+import { TransitionType } from '../../model/enums/transition-type'
+import { PartTimings } from '../../model/value-objects/part-timings'
+import { PieceLifespan } from '../../model/enums/piece-lifespan'
+import { UnsupportedOperation } from '../../model/exceptions/unsupported-operation'
+
+const ACTIVE_GROUP_PREFIX: string = 'active_group_'
+const PREVIOUS_GROUP_PREFIX: string = 'previous_group_'
+const NEXT_GROUP_PREFIX: string = 'next_group_'
+const INFINITE_GROUP_PREFIX: string = 'infinite_group_'
+const PIECE_PRE_ROLL_PREFIX: string = 'pre_roll_'
+
+const PIECE_CONTROL_INFIX: string = '_piece_control_'
+const PIECE_GROUP_INFIX: string = '_piece_group_'
+
+// These priority values are the same values used by Core
+const HIGH_PRIORITY: number = 5
+const MEDIUM_PRIORITY: number = 1
+const LOW_PRIORITY: number = -1
+
+export class SuperflyTimelineBuilder implements TimelineBuilder {
+	public getBaseTimeline(): Timeline {
+		return { timelineGroups: [] }
+	}
+
+	public buildTimeline(rundown: Rundown): Timeline {
+		const timeline: Timeline = {
+			timelineGroups: [],
+		}
+
+		// TODO: Find LookAhead
+		// TODO: Find Baseline
+
+		const activePartTimelineGroup: ActivePartTimelineObjectGroup =
+			this.createTimelineObjectGroupForActivePart(rundown)
+		timeline.timelineGroups.push(activePartTimelineGroup)
+
+		const previousPart: Part | undefined = rundown.getPreviousPart()
+		if (previousPart) {
+			const previousPartTimelineGroup: TimelineObjectGroup = this.createTimelineObjectGroupForPreviousPart(
+				previousPart,
+				rundown.getActivePart(),
+				activePartTimelineGroup
+			)
+			timeline.timelineGroups.push(previousPartTimelineGroup)
+		}
+
+		const infinitePiecesTimelineGroups: TimelineObjectGroup[] =
+			this.createTimelineObjectGroupsForInfinitePieces(rundown)
+		timeline.timelineGroups.push(...infinitePiecesTimelineGroups)
+
+		if (activePartTimelineGroup.shouldAutoNextAtPointInTime > 0) {
+			const nextPartTimeline: TimelineObjectGroup = this.createTimelineForNextPart(
+				rundown,
+				activePartTimelineGroup
+			)
+			timeline.timelineGroups.push(nextPartTimeline)
+
+			if (Number.isNaN(activePartTimelineGroup.enable.duration)) {
+				throw new UnsupportedOperation(
+					`Duration of activePartTimelineGroup must be a number! Got: ${activePartTimelineGroup.enable.duration}`
+				)
+			}
+			timeline.autoNext = {
+				pointInTimeToTakeNext:
+					activePartTimelineGroup.shouldAutoNextAtPointInTime -
+					rundown.getNextPart().getTimings().previousPartContinueIntoPartDuration,
+			}
+		}
+
+		// TODO: Call Blueprint "onTimelineGenerate". This will most likely need some tweaks.
+
+		return timeline
+	}
+
+	private createTimelineObjectGroupForActivePart(rundown: Rundown): ActivePartTimelineObjectGroup {
+		const now: number = Date.now()
+
+		const activePart: Part = rundown.getActivePart()
+
+		const currentPartEnable: TimelineEnable = {
+			start: now,
+		}
+
+		let shouldAutoNextAtPointInTime: number = 0
```

Comment:
As mentioned for the `buildTimeline` method of this class, the `if` could be pulled to its own method.

Might be overkill to do so here. However doing it, you would allow you to change `let shouldAutoNextAtPointInTime: number` to `const shouldAutoNextAtPointInTime: number`, as the result of the method would be assigned directly, instead of conditionally.

Reply:
How would you implement that method without making it have "side effects" and not returning weird "types" like `{ timelineEnable: TimelineEnable, shouldAutoNextAtPointInTime: number }`?

Reply:
Do you mean that assinging to `currentPartEnable.duration` is an "side effects"?

If that is the case, and you do not want to do that, then i would be fine with keeping this `if` statements, as it isn't that big, compared to some of the others.